src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java

```diff
@@ -194,10 +198,21 @@ public void removeAclOnFiles(Dataset dataset, List<String> fileIds, Map<IamRole,
         fileAclOp(AclOp.ACL_OP_DELETE, dataset, fileIds, policies);
     }
 
-    private static final String GS_PROTOCOL = "gs://";
-    private static final String GS_BUCKET_PATTERN = "[a-z0-9_.\\-]{3,222}";
-
     public static Blob getBlobFromGsPath(Storage storage, String gspath, String targetProjectId) {
+        GcsLocator locator = getGcsLocatorFromGsPath(gspath);
+
+        // Provide the project of the destination of the file copy to pay if the
+        // source bucket is requester pays.
+        Blob sourceBlob = storage.get(BlobId.of(locator.getBucket(), locator.getPath()),
+            Storage.BlobGetOption.userProject(targetProjectId));
+        if (sourceBlob == null) {
+            throw new PdaoSourceFileNotFoundException("Source file not found: '" + gspath + "'");
+        }
+
+        return sourceBlob;
+    }
+
+    public static GcsLocator getGcsLocatorFromGsPath(String gspath) {
```

Comment:
Since this is inside a class called `GcsPdao` IMO it would just as clear to call this `getLocatorFromPath()` since any uses of it will have to have the `GcsPdao` prefix.